p2p/pex_reactor.go

```diff
@@ -293,6 +309,160 @@ func (r *PEXReactor) ensurePeers() {
 	}
 }
 
+// Explores the network searching for more peers. (continuous)
+// Seed/Crawler Mode causes this node to quickly disconnect
+// from peers, except other seed nodes.
+func (r *PEXReactor) seedCrawlerMode() {
+	// Do an initial crawl
+	r.crawlPeers()
+
+	// Fire periodically
+	ticker := time.NewTicker(5 * time.Second)
+
+	for {
+		select {
+		case <-ticker.C:
+			r.attemptDisconnects()
+			r.crawlPeers()
+		case <-r.Quit:
+			return
+		}
+	}
+}
+
+// crawlStatus handles temporary data needed for the
+// network crawling performed during seed/crawler mode.
+type crawlStatus struct {
+	// The remote address of a potential peer we learned about
+	Addr *NetAddress
+
+	// Not empty if we are connected to the address
+	PeerID string
+
+	// The last time we attempt to reach this address
+	LastAttempt time.Time
+
+	// The last time we successfully reached this address
+	LastSuccess time.Time
+}
+
+// oldestAttempt implements sort.Interface for []crawlStatus
+// based on the LastAttempt field.
+type oldestAttempt []crawlStatus
+
+func (oa oldestAttempt) Len() int           { return len(oa) }
+func (oa oldestAttempt) Swap(i, j int)      { oa[i], oa[j] = oa[j], oa[i] }
+func (oa oldestAttempt) Less(i, j int) bool { return oa[i].LastAttempt.Before(oa[j].LastAttempt) }
+
+// getCrawlStatus returns addresses of potential peers that we wish to validate.
+// NOTE: The status information is ordered as described above.
+func (r *PEXReactor) getCrawlStatus() []crawlStatus {
+	var oa oldestAttempt
+
+	addrs := r.book.List()
+	// Go through all the addresses in the AddressBook
+	for _, addr := range addrs {
+		oa = append(oa, crawlStatus{
+			Addr:        addr.Addr,
+			PeerID:      r.addrToPeerKey(addr.Addr),
+			LastAttempt: addr.LastAttempt,
+			LastSuccess: addr.LastSuccess,
+		})
+	}
+	sort.Sort(oa)
+	return oa
+}
+
+// addrToPeerKey returns the Peer.Key() if the Peer is currently
+// connected and has the RemoteAddr equal to the addr parameter.
+func (r *PEXReactor) addrToPeerKey(addr *NetAddress) string {
+	var id string
+
+	peers := r.Switch.peers.List()
+	for _, p := range peers {
```

Comment:
there is a `peers.Has` function

Reply:
Yes, but it performs the lookup using the peer.Key(), not the peer.NodeInfo().RemoteAddr

Reply:
Maybe we need to rename `Has` to `FindByID()` and add `FindByRemoteAddr()`?

Reply:
Done.